falcon/cluster/cluster.py

```diff
@@ -145,93 +187,138 @@ def _build_query_ann_index(
-        with open(pkl_filename, 'rb') as f_in:
-            for spec_raw in pickle.load(f_in):
-                spec_processed = process_spectrum(spec_raw)
-                # Discard low-quality spectra.
-                if spec_processed is not None:
-                    spectra_split.append(spec_processed)
-                    filenames.append(spec_processed.filename)
-                    identifiers.append(spec_processed.identifier)
-                    precursor_mzs_split.append(spec_processed.precursor_mz)
-                    rts_split.append(spec_processed.retention_time)
-        if len(spectra_split) == 0:
-            continue
-        precursor_mzs.append(np.asarray(precursor_mzs_split))
-        rts.append(np.asarray(rts_split))
-        # Convert the spectra to vectors.
-        vectors_split = vectorize(spectra=spectra_split)
-        n_split, dim = vectors_split.shape
-        # Figure out a decent value for the n_list hyperparameter based on
-        # the number of vectors.
-        # Rules of thumb from the Faiss wiki:
-        # https://github.com/facebookresearch/faiss/wiki/Guidelines-to-choose-an-index#how-big-is-the-dataset
-        if n_split == 0:
-            continue
-        if n_split < 10**2:
-            # Use a brute-force index instead of an ANN index when there
-            # are only a few items.
-            n_list = -1
-        elif n_split < 10**6:
-            n_list = 2**math.floor(math.log2(n_split / 39))
-        elif n_split < 10**7:
-            n_list = 2**16
-        elif n_split < 10**8:
-            n_list = 2**18
-        else:
-            n_list = 2**20
-            if n_split > 10**9:
-                logger.warning('More than 1B vectors to be indexed, consider '
-                               'decreasing the ANN size')
-        # Create an ANN index using the inner product (proxy for cosine
-        # distance) for fast NN queries.
-        if n_list <= 0:
-            index = faiss.IndexIDMap(faiss.IndexFlatIP(dim))
-        else:
-            index = faiss.IndexIVFFlat(faiss.IndexFlatIP(dim), dim, n_list,
-                                       faiss.METRIC_INNER_PRODUCT)
-            index.nprobe = min(math.ceil(index.nlist / 8), n_probe)
-        # Compute cluster centroids.
+    msms_spectra = spectra_df.apply(spectrum.df_row_to_spec, axis=1).tolist()
+    for spec_raw in msms_spectra:
+        spec_processed = process_spectrum(spec_raw)
+        # Discard low-quality spectra.
+        if spec_processed is not None:
+            spectra.append(spec_processed)
+            filenames.append(spec_processed.filename)
+            identifiers.append(spec_processed.identifier)
+            bucket_ids.append(bucket_id)
+            precursor_mzs.append(spec_processed.precursor_mz)
+            rts.append(spec_processed.retention_time)
+    if len(spectra) == 0:
+        logger.warning("No spectra to cluster after quality inspection")
```

Comment:
**todo:** We could also log an info message telling the user how many spectra were discarded after quality filtering (if any).